README.md

```diff
@@ -39,102 +80,286 @@ data Sample = Sample
+types the flag, and `False` otherwise.
+
+There are other kinds of basic parsers, and several ways to configure them.
+These are covered in the [Builders](#builders) section.
+
+### Applicative
+
+Now we may combine the `target` and `quiet` into a single parser that accepts
+both options and returns a combined value. Given a type
+
+```haskell
+data Options = Options
+  { optTarget :: String
+  , optQuiet :: Bool }
+```
+
+and now it's just a matter of using `Applicative`'s apply operator `(<*>)`
+to combine the two previously defined parsers
+
+```haskell
+opts :: Parser Options
+opts = Options <$> target <*> quiet
 ```
 
- [applicative]: http://www.soi.city.ac.uk/~ross/papers/Applicative.html
+No matter which parsers appear first in the sequence, options will still be
+parsed in whatever order they appear in the command line. A parser with such a
+property is sometimes called a *permutation parser*.
+
+In our example, a command line like:
+
+    --target world -q
+
+will give the same result as
+
+    -q --target world
+
+It is this property which leads us to an Applicative interface instead of a
+Monadic one, as all option must be considered in parallel, and can not depend
+on the output of other options.
+
+Note, however, that the order of sequencing is still somewhat significant, in
+that it affects the generated help text. Customisation can be achieved easily
+through a lambda abstraction, with [Arrow notation](#arrow-interface), or by
+taking advantage of GHC 8's [ApplicativeDo](#applicative-do) extension.
+
+### Alternative
 
-## Supported options
+It is also common to find programs that can be configured in different ways
+through the command line.  A typical example is a program that can be given a
+text file as input, or alternatively read it directly from the standard input.
 
-`optparse-applicative` supports four kinds of options: regular options, flags,
-arguments and commands.
+We can model this easily and effectively in Haskell using *sum types*:
+
+```haskell
+data Input
+  = FileInput FilePath
+  | StdInput
+
+run :: Input -> IO ()
+run = ...
+```
+
+We can now define two basic parsers for the components of the sum type:
+
+```haskell
+fileInput :: Parser Input
+fileInput = FileInput <$> strOption
+  (  long "file"
+  <> short 'f'
+  <> metavar "FILENAME"
+  <> help "Input file" )
+
+stdInput :: Parser Input
+stdInput = flag' StdInput
+  (  long "stdin"
+  <> help "Read from stdin" )
+```
+
+As the `Parser` type constructor is an instance of `Alternative`, we can compose
+these parsers with a choice operator `(<|>)`
+
+```haskell
+input :: Parser Input
+input = fileInput <|> stdInput
+```
+
+Now `--file "foo.txt"` will be parsed as `FileInput "foo.txt"`, `--stdin` will
+be parsed as `StdInput`, but a command line containing both options, like
+
+    --file "foo.txt" --stdin
+
+will be rejected.
+
+Having `Applicative` and `Alternative` instances, optparse-applicative `Parser`s
```

Comment:
s/`Parser`s/Parsers or `Parser`'s

doesn't really make sense to have it quoted here